src/biome/allennlp/dataset_readers/sequence_classifier_dataset_reader.py

```diff
@@ -65,61 +66,85 @@ def _read(self, file_path: str) -> Iterable[Instance]:
-            file_path, SequenceClassifierForwardConfiguration
-        )
+        data_source = DataSource.from_yaml(file_path)
 
-        ds_key = file_path
-        dataset = self.get(ds_key)
-        if dataset is not None:
-            logger.debug(f"Loaded cached dataset {file_path}")
-            return dataset
+        # get cached instances of the data set
+        instances = self.get(file_path)
+        if instances is not None:
+            logger.debug("Loaded cached data set {}".format(file_path))
         else:
-            logger.debug(f"Read dataset from {file_path}")
-            dataset = self._read_as_forward_dataset(data_source, forward)
-            instances = dataset[["tokens", "label"]].apply(
-                self.text_to_instance, axis=1
-            )
-            self.set(ds_key, instances)
-            return (instance for idx, instance in instances.iteritems() if instance)
-
-    @staticmethod
-    def _read_as_forward_dataset(
-        data_source: DataSource, forward: SequenceClassifierForwardConfiguration
-    ):
-        dataset = data_source.to_dataframe().compute()
-
-        dataset["tokens"] = dataset[forward.tokens].apply(lambda x: x.to_dict(), axis=1)
-        dataset["label"] = (
-            dataset[forward.label].astype(str).apply(forward.sanitize_label)
-        )
+            logger.debug("Read data set from {}".format(file_path))
+            dataset = data_source.to_forward_dataframe()
+            instances = dataset.apply(self.text_to_instance, axis=1)
 
-        return dataset
+            # cache instances of the data set
+            self.set(file_path, instances)
+
+            # If memory is an issue maybe we should only cache the dataset and yield instances:
+            # for example in dataset.itertuples(index=False):
+            #     yield self.text_to_instance(example)
+
+        return (instance for idx, instance in instances.iteritems() if instance)
 
-    @overrides
     def text_to_instance(
-        self, example: Dict[str, Union[Dict[str, str], str]]
+        self, example: Dict[str, str], exclude_optional: bool = False
     ) -> Optional[Instance]:
         """Extracts the forward parameters from the example and transforms them to an `Instance`
 
         Parameters
         ----------
         example
             The keys of this dictionary should match the arguments of the `forward` method of your model.
+        exclude_optional
+            Only extract the mandatory parameters of the model's forward method.
 
         Returns
         -------
         instance
             Returns `None` if the example could not be transformed to an Instance.
         """
+        fields = {}
+        for param_name, param in self.forward_params.items():
+            if param_name == "self":
+                continue
+            # if desired, skip optional parameters, like the label for example
+            if exclude_optional and param.default is not Parameter.empty:
```

Comment:
I think including the `exclude_optional` parameter can bring problems on prediction time. If i understand well the purpose of this flag, when model is serving for inference, the `Predictor` should call the reader excluding default paramas:

```python
#...predictor code
self._reader.text_to_instance(data, exclude_optional=True)
#...more predictor code
```
while in the `predict` command, the predictor should call `text_to_instance` disabling `exclude_optional` flag, if not, we'll break the `predict` feature

This behaviour is not easy to implement using the same predictor because the allennlp api

I propose evaluate dynamically the data example values and apply the `_value_to_field` method just if the param value is defined. 

This approach brings interesting features, like check model metrics in serving inference time (providing a input with a `label` attribute)

Reply:
Very true! I removed this parameter for the moment.
In a separate PR this week i will address the issues with the predict and serve cases!